### DISCUSSION TITLE
Use groups from OpenID Connect SSO provider

### DISCUSSION THREAD

**Discussion by Frando** [outline/outline#3779] [Ideas]
Hi,
we're using Outline with Keycloak as an OpenID Connect (OIDC) single-sign on provider. This works great.
In Keycloak, we maintain group membership for our users. If I see things correctly, it is currently not possible to use these groups in Outline.

This would be great to have. We'd like to split our Outline wiki in different sections, accessible to different parts of the member audience. This is possible in a nice and simple way by setting up groups in Outline and then setting the per-collection permissions accordingly.

However, we would love to not have to do this manually, as all the required groups and group memberships are already maintained in our Keycloak user store. For example Nextcloud has a feature (through the social login extension) that allows to map OIDC groups (exposed in a groups claim) to Nextcloud group, or optionally auto-create Nextcloud groups for all groups i a OIDC claim that do not yet exist.

This feature would be great to have in Outline, as it would allow us to have a single source of truth for group memberships and not have to maintain the group memberships in outline seperately.

Thanks!
Frando

**Comment by Frando** [comment]
I created a pull request for this feature: #3785

**Reply by almereyda** [reply]
Thank you for this example implementation!

We have seen preliminary work in this direction after https://github.com/outline/outline/pull/2555#issuecomment-934412345 and #2568, which ultimately stalled.

I will be keen to test this PR against the GitLab OIDC IDP, and especially like the idea of decoupling the login flow related implementation into the `userGroupUpdater`.

**Reply by lexermal** [reply]
How does the progress look like?

**Reply by 16flotho** [reply]
Waiting too...

**Reply by thomotron** [reply]
Unfortunately this is the one thing holding us back from implementing Outline in our stack. Keen to see this merged.

**Comment by routerino** [comment]
I could have sworn there is a way to assign groups via SSO, but I'm not finding it now.

Perhaps Outline could implement [SCIM](https://simplecloud.info/) instead? Given that SCIM is API/JSON driven, that's probably not a large ask, and most major providers (including keycloak, albeit via plugin) supports SCIM.

**Comment by Frando** [comment]
Hi all, after having no tangible path forward for #3785 I chose a different path and wrote a small webhook server with Deno. It subscribes to `users.signin` events from Outline, and syncs up their groups from Keycloak via the Outline and Keycloak REST APIs. It works well.

I just open-sourced it here: https://gist.github.com/Frando/aa561ca7e6c72ab64b5d17df911c0b1f
Feel invited to try it out and report feedback here or on the gist.

**Reply by Altavion** [reply]
Here's one more script to sync groups from an identity provider to Outline which works with Authentik instead of Keycloak:

https://gist.github.com/Altavion/45565e0635d20fa3860f3bbe9969b4bc

**Reply by Altavion** [reply]
And another take on the same problem:

https://github.com/burritosoftware/Outline-Authentik-Connector

**Comment by acul021** [comment]
In case anyone still needs this: we ended up solving it via SCIM instead of OIDC claims. I wrote a small sidecar that translates SCIM 2.0 to Outline's REST API, so groups sync from the IdP continuously instead of only at login: https://github.com/acul021/scim-outline-adapter (more details in https://github.com/outline/outline/discussions/12984)
